File: typescript/src/pages/heroku/broken-images.page.ts
import { BasePage } from '../base.page';

/** Broken Images page (/broken_images). */
export class BrokenImagesPage extends BasePage {
  /**
   * Rendered width of an image; broken images render at 0px (mirrors the
   * Cypress naturalWidth checks).
   */
  async naturalWidth(src: string): Promise<number> {
    const image = this.page.locator(`img[src='${src}']`);
    await image.waitFor();
    return image.evaluate((el: HTMLImageElement) => el.naturalWidth);
  }

  /** Wait until the image has finished loading (only for images expected to be valid). */
  async waitForLoaded(src: string): Promise<void> {
    await this.page.waitForFunction(
      (s: string) => {
        const img = document.querySelector(`img[src='${s}']`) as HTMLImageElement | null;
        return !!img && img.complete && img.naturalWidth > 0;
      },
      src,
      { timeout: 10_000 },
    );
  }
}

File: src/main/java/com/example/ui/pages/heroku/BrokenImagesPage.java
package com.example.ui.pages.herokuapp;

import com.example.base.TestContext;
import com.example.ui.pages.BasePage;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import lombok.extern.slf4j.Slf4j;

/**
 * Broken Images page ({@code /broken_images}).
 */
@Slf4j
public class BrokenImagesPage extends BasePage {

    public BrokenImagesPage(Page page, TestContext testContext) {
        super(page, testContext);
    }

    /**
     * Returns the rendered width of an image; broken images render at 0px
     * (mirrors the Cypress {@code naturalWidth} checks).
     */
    public int naturalWidth(String src) {
        Locator image = page.locator("img[src='" + src + "']");
        image.waitFor();
        Object width = image.evaluate("el => el.naturalWidth");
        int result = ((Number) width).intValue();
        log.debug("Image '{}' naturalWidth={}", src, result);
        return result;
    }
}
